### PR TITLE
Fix: `no-extra-bind` false positive (fixes #5058)

### LINKED ISSUE #5058: Bug in no-extra-bind
I found another bug in this rule.

If the `this` keyword only appears as the parameter to `bind`, it doesn't get registered in the right scope because we've already exited the `FunctionExpression`, but not yet from its associated `CallExpression`.

So in code like this:

```javascript
var a = function() {
  (function() {
    this.b
  }.bind(this))
}.bind(c)
```

the outer `bind(c)` is marked as unnecessary, even though `this` is actually used in the bound function.

The bug gets fixed (and other tests will also pass) if I replace this:

```javascript
if (top.depth === 0) {
    top.found++;
}
```

with this:

```javascript
if (top.depth === 0) {
    top.found++;
} else if (top.depth === -1 && scope.length > 1) {
    var parentOfTop = scope[scope.length - 2];
    parentOfTop.found++;
}
```

It feels slightly hacky, but I'm not sure how to solve it in a cleaner way.

I can submit a PR with this fix if needed.

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/30163329-bug-in-no-extra-bind?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by mysticatea**
Fixes #5058.

### CODE REVIEW

File: tests/lib/rules/no-extra-bind.js [hunk 1]
```diff
@@ -33,6 +36,6 @@ ruleTester.run("no-extra-bind", rule, {
         { code: "var a = (() => { return this; }).bind(b)", parserOptions: { ecmaVersion: 6 }, errors: [{ message: "The function binding is unnecessary.", type: "CallExpression"}] },
         { code: "var a = function() { (function(){ this.c }) }.bind(b)", errors: [{ message: "The function binding is unnecessary.", type: "CallExpression"}] },
         { code: "var a = function() { function c(){ this.d } }.bind(b)", errors: [{ message: "The function binding is unnecessary.", type: "CallExpression"}] },
-        { code: "var a = function() { (function(){ (function(){ this.d }.bind(c)) }) }.bind(b)", errors: [{ message: "The function binding is unnecessary.", type: "CallExpression", column: 9}] }
+        { code: "var a = function() { (function(){ (function(){ this.d }.bind(c)) }) }.bind(b)", errors: [{ message: "The function binding is unnecessary.", type: "CallExpression", column: 71}] }
```
Comment (mysticatea) [inline tests/lib/rules/no-extra-bind.js]: I moved the location from the head of function to the `.bind()` method.

File: tests/lib/rules/no-extra-bind.js [hunk 2]
```diff
@@ -25,7 +25,10 @@ ruleTester.run("no-extra-bind", rule, {
         "var a = function() { this.b }.foo()",
         "var a = f.bind(a)",
         "var a = function() { return this.b }.bind(c)",
-        { code: "var a = (() => { return b }).bind(c, d)", parserOptions: { ecmaVersion: 6 } }
+        { code: "var a = (() => { return b }).bind(c, d)", parserOptions: { ecmaVersion: 6 } },
+        "(function() { (function() { this.b }.bind(this)) }.bind(c))",
+        "var a = function() { return 1; }[bind](b)",
```
Comment (mjomble) [inline tests/lib/rules/no-extra-bind.js]: Shouldn't it be `['bind']` instead of `[bind]`?

Comment (mysticatea) [inline tests/lib/rules/no-extra-bind.js]: Thank you for reviewing!

This test verifies this rule ignores computed properties.

```js
var bind = "foo";
var a = function() { return 1; }[bind](b); // this shouldn't be warned!
```

The function does not have `this` keywords, if it's `"bind"` then this should be warned.
....... but not warned currently.
I will fix it.

Comment (mysticatea) [inline tests/lib/rules/no-extra-bind.js]: Done.

**Comment by nzakas** [comment]
Lgtm

**Pull request merged**